ministryofjustice/laa-apply-for-criminal-legal-aid
# [CRIMAPP-614/622] Normalise housing payment form behaviours, Update council tax rules

## Description of change
Updates housing type and council tax forms to prevent inconsistent data entry

## Link to relevant ticket
Data consistency: https://dsdmoj.atlassian.net/browse/CRIMAPP-614
Board and Lodging no council tax: https://dsdmoj.atlassian.net/browse/CRIMAPP-622

## Notes for reviewer
- Updates 'None' housing type wording for summary
- Ensures if Board and Lodging selected then council tax entry is deleted
- Ensures user is redirected to housing payment type form if going directly to /rent_payments /mortgage_payments /board_and_lodging_payments
- Changes `outgoings_payments` db table back to non-nullable values to ensure Summary output and submissions are straightforward (i.e. payments are always correct)
- Ensures consistency with `income_benefits` and `income_payments` db tables, ready for CRIMAP-565 payments unification (reintroduces not null constraints to outgoings_payments)
- When completing housing payment type form removes any older housing payment entries if required
- Note alteration to `spec/support/shared_examples/step_controller_shared_examples.rb` which allows overriding of the `existing_case` variable by the calling context
- Fixes `PaymentFrequencyType.to_phrase` when called with a null value which resulted in funny output (should not be possible to do this in the Summary page because the frequency is now always required by the db table)
- Outstanding is the way HousingPaymentType and OutgoingsPaymentType are used interchangeably due to the way the 2 types were created/used throughout - requires further consideration
- If the user goes to /does_client_pay_council_tax directly with a board_and_lodging housing payment type already selected, then an error message is shown (wording requires consideration)
- Includes bug fix to payment pages allowing the user to enter no options without ticking 'none' first

## Screenshots of changes (if applicable)
No change to output other than the 'None' wording being changed. The changes are User Journey and business rules

### Before changes:
<img width="574" alt="Screenshot 2024-03-14 at 01 08 06" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/47113046/63df07e6-8b5b-4fec-a946-84abc9726710">

### After changes:
<img width="570" alt="Screenshot 2024-03-14 at 01 07 08" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/47113046/3a637122-3ddd-4ee2-a6ca-dcc5b95db66e">


Error message on council tax form if user goes to the page directly via URL



<img width="588" alt="Screenshot 2024-03-14 at 11 02 51" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/47113046/548151e1-1440-4f00-8a59-2d1d5e1bf1d9">

Error messages shown when no income or outgoing payment type is selected:

<img width="599" alt="Screenshot 2024-03-18 at 10 45 48" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/47113046/1fb055c6-68e6-48c5-8cb0-d3b7e21dafcd">
<img width="600" alt="Screenshot 2024-03-18 at 10 47 59" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/47113046/3ea87630-d962-4fb2-b195-902d02f115b1">
<img width="599" alt="Screenshot 2024-03-18 at 10 48 19" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/47113046/c0b6cffa-50d7-4b66-a42c-5a502a080d1a">




## How to manually test the feature
Create an application and go to <id>/steps/outgoings/rent_payments and you should be redirected to the housing payment type form and continue testing:

- add rent payments
- add council tax payments
- go  back to the housing payment type form and change to board and lodging
- should not show council tax form
- go to application review summary page and there should not be any council tax information only board and lodging
